Assignments/ip2/ip2.md

```diff
@@ -0,0 +1,310 @@
+---
+layout: assignment
+title: "Individual Project 2"
+permalink: /assignments/ip2
+parent: Assignments
+nav_order: 1
+due_date: "Wednesday October 18, 11:00am EDT"
+submission_notes: Submit via Autograder.io at neu.autograder.io
+---
+
+The objective for this semester's individual project is to implement this new `GameArea` abstraction, with one concrete implementation: `TicTacToeArea`. 
+The `TicTacToeArea` will implement the gameplay for the classic game, [Tic-Tac-Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).
+This implementation effort will be split across two deliverables. In this second deliverable, you will implement and (partially) test the core frontend components. 
+
+
+## Objectives of this assignment
+The objectives of this assignment are to:
+* Investigate and understand a large, existing codebase
+* Write new TypeScript code that uses asynchronous operations
+* Write test cases that utilize mocks and spies
+* Write React components and hooks that make use of state
+
+
+## Getting started with this assignment
+Start by [downloading the starter code]({{site.baseurl}}{% link /Assignments/ip2/ip2-handout.zip %}). Extract the archive and run `npm install` to fetch the dependencies. 
+
+**Configuring Jest and VSCode**: If you would like to use the built-in Jest test runner for VSCode (where it shows the tests and their status in the sidebar), the easiest way to accomplish this for this project is to open *just* the "frontend" directory or just the "townService" directory in VSCode - not the top-level "ip2-handout" directory. If you have a quick-fix to make it work with the whole project at once, please feel free to share on Piazza and we will incorportate that here.
```

Comment:
I was able to configure my VSC to treat `frontend` and `townService` as two separate projects by using [multi-root workspaces](https://code.visualstudio.com/docs/editor/workspaces#_multiroot-workspaces). All the extensions, including ESLint, Prettier, and Jest seem to work fine. Here's what I did:

1. Create a new `ip2-handhout.code-workspace` file in the root level `ip2-handout` directory. This should notify VSC to treat the project as 3 different projects.
```json
{
	"folders": [
		{
			"path": ".",			
		}, 
		{
			"path": "./frontend",
		},
		{
			"path": "./townService"
		}
	],
	"settings": {
		"typescript.tsdk": "node_modules/typescript/lib",
	}
```
2. Edit `.vscode/settings.json` to include the following lines:
```json
"eslint.workingDirectories": [
        "./frontend",
        "./townService",
 ]
```
3. Create a new file, `.vscode/tasks.json` as follows. This provides us with a handy tool to start backend and frontend development servers without manually `cd`-ing into those folders.
```json
{
  "version": "2.0.0",
  "tasks": [
    {
      "label": "Start React Development Server",
      "type": "npm",
      "script": "dev",
      "path": "frontend",
      "group": {
        "kind": "build",
        "isDefault": true
      }
    },
    {
      "label": "Start townService",
      "type": "npm",
      "script": "start",
      "path": "townService",
      "group": {
        "kind": "build",
        "isDefault": true
      }
    }
  ]
}
```
Now, press `Ctrl + Shift + P` (or `Cmnd + Shift + P`) > Search for "Run Tasks" > Select the appropriate task from the dropdown to start React or Node dev servers. \

I can edit the `ip2-handout.zip` file with the updated configurations if you'd like.

Reply:
I get this when I try that:
```**[0] [nodemon] starting `ts-node src/Server.ts`
[0] TypeError [ERR_UNKNOWN_FILE_EXTENSION]: Unknown file extension ".ts" for /Users/jon/Documents/NEU/Teaching/covey.town-private/townService/src/Server.ts
[0]     at new NodeError (node:internal/errors:405:5)
[0]     at Object.getFileProtocolModuleFormat [as file:] (node:internal/modules/esm/get_format:99:9)
[0]     at defaultGetFormat (node:internal/modules/esm/get_format:142:36)
[0]     at defaultLoad (node:internal/modules/esm/load:115:20)
[0]     at nextLoad (node:internal/modules/esm/hooks:832:28)
[0]     at load (/Users/jon/Documents/NEU/Teaching/covey.town-private/townService/node_modules/ts-node/dist/child/child-loader.js:19:122)
[0]     at nextLoad (node:internal/modules/esm/hooks:832:28)
[0]     at Hooks.load (node:internal/modules/esm/hooks:415:26)
[0]     at MessagePort.handleMessage (node:internal/modules/esm/worker:168:24)
[0]     at [nodejs.internal.kHybridDispatch] (node:internal/event_target:807:20)
```

I am going to include the configuration files, but will hold off on recommending this until we get more feedback (would rather avoid 150 students finding out that it doesn't work and all writing in about it!)